growth/too/tasks/ref.py

```diff
@@ -0,0 +1,71 @@
+import sys
+import subprocess
+import shlex
+from subprocess import PIPE
+import numpy as np
+import tempfile
+import requests
+
+from celery.task import PeriodicTask
+from celery.utils.log import get_task_logger
+
+from ..flask import app
+from . import celery
+from .. import models
+
+log = get_task_logger(__name__)
+
+__all__ = ('ztf_references',)
+
+
+@app.task(base=PeriodicTask, shared=False, run_every=3600)
+def ztf_references():
+
+    user = app.config['IRSA_USERNAME']
+    pw = app.config['IRSA_PASSWORD']
+
+    if not 'IRSA_USERNAME' in app.config or not 'IRSA_PASSWORD' in app.config:
+        log.info('Missing IRSA_USERNAME or IRSA_PASSWORD...')
+        return
+
+    params = (('josso_cmd', 'login'),
+              ('josso_username', app.config['IRSA_USERNAME']),
+              ('josso_password', app.config['IRSA_PASSWORD']))
+
+    session = requests.Session()
+    session.get('https://irsa.ipac.caltech.edu/account/signon/login.do', params=params, verify=False)
+    params = (('WHERE', 'nframes>=15 AND startobsdate>=\'2018-02-05T00:00:00Z\''),)
+    r = session.get('https://irsa.ipac.caltech.edu/ibe/search/ztf/products/ref', params=params, verify=False)
+
+    lines = r.text.split('\n')
+    refs = {}
+    mags = {}
+    for line in lines:
+        lineSplit = line.split(" ")
+        lineSplit = list(filter(None,lineSplit))
+    
+        if not len(lineSplit) == 35: continue
+    
+        fieldID, filt, maglimcat = int(lineSplit[1]), int(lineSplit[5]), float(lineSplit[9])
+        if not fieldID in refs:
+            refs[fieldID] = []
+            mags[fieldID] = []
+        refs[fieldID].append(filt)
+        mags[fieldID].append(maglimcat)
+    for fieldID in refs.keys():
+        refs[fieldID], idx = np.unique(refs[fieldID],return_index=True)
+        mags[fieldID] = np.array(mags[fieldID])[idx]
+        
+    if len(refs.keys()) == 0:
+        log.info('References missing... quitting.')
+        return
+
+    fields = models.Field.query.filter_by(telescope='ZTF').all()
```

Comment:
You don't have to query the database first. You can create new instances of `models.Field`, fill in the values that you want to change, and add them to the session with `models.db.session.merge()`. For objects that already exist in the database, that will map to SQL `UPDATE` statements.

Reply:
When i did that, it complained that ra and dec cannot be null.

Reply:
Please show me the code that you used to do that.

Reply:
I just wrote something like:
models.Field(field_id = field_id, ref_filter_ids=ref_filter_ids, ref_filter_mags=ref_filter_mags)
or similar.

Reply:
Dd you set telescope? You probably have to specify all of the values that belong to the primary key.